fix: filter by version tag in search process definition endpoint

Rsousa3 opened the pull request:
## Description

This PR adds a version tag filter to the `/v1/process-definitions/search` endpoint.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)
  - [ ] enable backports [when recommended](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)

## Related issues

closes #30374

Rsousa3 committed:
fix: added version tag filters to search process definition

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda?pullRequest=30398) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda?pullRequest=30398) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda?pullRequest=30398) it.</sub>

kristinkomschow reviewed (changes requested):
Thank you for contributing to this!

There are some smaller suggestions in the code base and two more things:

There is still an integration test missing for checking if the filtering returns the expected results. This wasn't mentioned in the issue, but should be added to [ProcessDefinitionDaoIT](https://github.com/camunda/camunda/blob/main/operate/qa/integration-tests/src/test/java/io/camunda/operate/webapp/api/v1/dao/ProcessDefinitionDaoIT.java).

I also saw that the first commit is a merge with main. Please use rebase instead as per our [contribution guidelines](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#updating-a-pull-request) because we want to keep our history clear of merge commits.

kristinkomschow commented:
> Hello @kristinkomschow. Thank you for your reply and comments! I will apply those changes, as well as use rebase next time. I will have an update on this PR tomorrow.

Thanks for the update!

Please also remove the existing merge commit (our CI might actually prevent us from merging the PR otherwise) and squash the upcoming changes.

Rsousa3 commented on operate/webapp/src/test/java/io/camunda/operate/webapp/api/v1/dao/opensearch/OpensearchProcessDefinitionDaoTest.java:
```diff
@@ -0,0 +1,73 @@
+        .setTenantId("fakeTenant")
+        .setName("fakeName")
+        .setVersion(1)
+        .setVersion(2);
```
Fixed! Can you have a look?

Rsousa3 commented on operate/webapp/src/test/java/io/camunda/operate/webapp/api/v1/dao/elasticsearch/ElasticsearchProcessDefinitionDaoTest.java:
```diff
@@ -0,0 +1,59 @@
+        .setKey(123L)
+        .setBpmnProcessId("fakeBpmnProcessId")
+        .setTenantId("fakeTenant")
+        .setName("fakeName")
```
Fixed! Can you have a look?

Rsousa3 commented on operate/webapp/src/test/java/io/camunda/operate/webapp/api/v1/dao/elasticsearch/ElasticsearchProcessDefinitionDaoTest.java:
```diff
@@ -0,0 +1,59 @@
+        .setTenantId("fakeTenant")
+        .setName("fakeName")
+        .setVersion(1)
+        .setVersion(2);
```
Fixed! Can you have a look?

Rsousa3 commented:
Hello @kristinkomschow. I just pushed the integration test `shouldFilterProcessDefinitionsByVersionTag`. Please let me know what you think and if anything else needs to be done.

kristinkomschow commented on operate/qa/integration-tests/src/test/java/io/camunda/operate/webapp/api/v1/dao/ProcessDefinitionDaoIT.java:
```diff
@@ -195,4 +198,15 @@ public void shouldPageProcessDefinitions() {
+        dao.search(
+            new Query<ProcessDefinition>()
+                .setFilter(new ProcessDefinition().setVersionTag("demoVersionTag")));
+
```
We should also validate here that there is only the one expected result, like:
`assertThat(processDefinitionResults.getItems()).hasSize(1);`

Rsousa3 replied:
Added the assertion!

kristinkomschow reviewed (changes requested):
Looks good to me apart from the last small thing I mentioned for the test!

I let the CI run and some of the checks are failing, but that seems to be because of https://github.com/camunda/camunda/issues/19222

Rsousa3 commented:
Thanks for the reply! I will add that assert to the integration test. About the naming in the other tests, I think I misunderstood at first. Could it instead be changed to "demoBpmnProcessId", "demoTenant", and so on, similar to the names on the integration tests? If so, I will make those changes and update the tests.

Rsousa3 commented:
Thank you for the clarification and the reminder. I have renamed the variables and squashed every commit into a single one. Sorry for the delay, and again, thank you for your patience and guidance!

brunofigalves commented on operate/webapp/src/test/java/io/camunda/operate/webapp/api/v1/dao/opensearch/OpensearchProcessDefinitionDaoTest.java:
```diff
@@ -0,0 +1,73 @@
+import io.camunda.operate.webapp.api.v1.entities.Query;
+import io.camunda.operate.webapp.opensearch.OpensearchQueryDSLWrapper;
+import io.camunda.operate.webapp.opensearch.OpensearchRequestDSLWrapper;
+import io.camunda.webapps.schema.descriptors.operate.index.ProcessIndex;
```
Hi @kristinkomschow, I guess @Rsousa3 had some issues running `mvn clean install` but I'll try to have a look 👀

kristinkomschow replied:
fyi: there are some dependencies in the operate code where you might get seemingly random compile issues. What I usually do is run `mvn clean install -Dquickly -PskipFrontendBuild` in the **project root directory**. Then in the Operate module run `mvn install -PskipFrontendBuild -DskipChecks -DskipQaBuild`.

kristinkomschow replied:
I also see this error when the tests are running:

> [ERROR]   ElasticsearchProcessDefinitionDaoTest.testBuildFilteringWithValidFields:51 NotAMock 
> Argument passed to verify() is of type ElasticsearchProcessDefinitionDao and is not a mock!
> Make sure you place the parenthesis correctly!
> See the examples of correct verifications:
>     verify(mock).someMethod();
>     verify(mock, times(10)).someMethod();
>     verify(mock, atLeastOnce()).someMethod();

Rsousa3 replied:
Hi @kristinkomschow. I was having problems running `mvn clean install`, like @brunofigalves said, but the commands you gave worked for me, so I was able to see the mentioned errors and fix them. After that, I was able to build in the Operate module, so I think there should be no more errors.

kristinkomschow reviewed (approved):
Looks good to me now!
I checked the CI failures but they don't stem from the changes here. Let's see if we can merge it despite them being there.

kristinkomschow commented:
Are you able to merge this or is it blocked (after signing the CLA)?
